File: menu-mini/src/api/dish.ts
import { request } from '@/utils/request'
import type { Page } from './common'

/** 菜谱（列表/详情共用；菜系/分类/标签名后端回填）。 */
export interface Dish {
  id: number
  name: string
  coverUrl?: string | null
  prepTime?: number | null
  cookTime?: number | null
  difficulty?: number | null
  note?: string | null
  sourceName?: string | null
  cuisineNames?: string[]
  categoryNames?: string[]
  tagNames?: string[]
  cookedCount?: number
}

/** 步骤（images 为逗号分隔的相对路径串）。 */
export interface DishStep {
  seq?: number
  text: string
  images?: string | null
}

/** 用料行（自然单位「2 个」「适量」）。 */
export interface DishIngredient {
  ingredientId: number
  ingredientName?: string | null
  amount?: number | null
  unitName?: string | null
}

export interface DishDetail {
  dish: Dish
  steps: DishStep[]
  ingredients: DishIngredient[]
}

/** 用量展示文本：「2 个」/「适量」/「2」。 */
export function amountText(ing: DishIngredient): string {
  const amt = ing.amount
  const num = amt == null ? '' : Number.isInteger(amt) ? String(amt) : String(amt)
  const unit = ing.unitName || ''
  if (!num && !unit) return ''
  if (!num) return unit
  return unit ? `${num} ${unit}` : num
}

export interface DishSearchParams {
  keyword?: string
  tagIds?: string // 逗号分隔多选
  cuisineIds?: string
  categoryIds?: string // 逗号分隔多选（2026-08-19 新标签体系：分类替代 tag+cuisine）
  sort?: 'cooked' | 'latest'
  pageNum: number
  pageSize?: number
}

/** 语义找菜结果项（向量相似召回）。 */
export interface SemanticHit {
  dishId: number
  name: string
  score?: number
  difficulty?: number | null
  cookTime?: number | null
}

/** 语义找菜：自然语言（「清淡下饭」「酸甜口」）→ 向量相似召回 TopK 菜谱。 */
export function semanticSearch(query: string, topK = 8): Promise<SemanticHit[]> {
  return request<SemanticHit[]>({
    url: '/dish/semantic-search',
    method: 'POST',
    data: { query, topK },
  })
}

/** 搜索菜谱：GET /dish/search（sort=cooked 做过最多，缺省最新）。 */
export function searchDishes(p: DishSearchParams): Promise<Page<Dish>> {
  return request<Page<Dish>>({
    url: '/dish/search',
    method: 'GET',
    data: {
      keyword: p.keyword || undefined,
      tagIds: p.tagIds || undefined,
      cuisineIds: p.cuisineIds || undefined,
      categoryIds: p.categoryIds || undefined,
      sort: p.sort === 'cooked' ? 'cooked' : undefined,
      pageNum: p.pageNum,
      pageSize: p.pageSize ?? 10,
    },
  })
}

/** 菜谱详情：GET /dish/{id}。 */
export function dishDetail(id: number): Promise<DishDetail> {
  return request<DishDetail>({ url: `/dish/${id}`, method: 'GET' })
}

/** 删除菜谱（连带步骤/关联/用料/历史）。 */
export function deleteDish(id: number): Promise<void> {
  return request({ url: `/dish/${id}`, method: 'DELETE' })
}

/** ===== 写菜谱（阶段 6） ===== */

export interface DishSavePayload {
  dish: {
    name: string
    coverUrl?: string
    note?: string
    prepTime?: number
    cookTime?: number
    difficulty?: number
  }
  steps: { seq?: number; sortOrder?: number; text: string; images?: string }[]
  ingredients: { ingredientId: number; amount?: number; unitId?: number }[]
  tagIds: number[]
  cuisineIds: number[]
}

/** 发布菜谱：POST /dish → 新菜 id。 */
export function saveDish(p: DishSavePayload): Promise<number> {
  return request<number>({ url: '/dish', method: 'POST', data: p })
}

/** 导入链接：POST /dish/import-url?url= → 新菜 id。 */
export function importDishByUrl(url: string): Promise<number> {
  return request<number>({ url: `/dish/import-url?url=${encodeURIComponent(url)}`, method: 'POST' })
}

export interface DishDraftItem {
  id: number
  name?: string | null
  coverUrl?: string | null
  ingredientCount?: number
  stepCount?: number
  updateTime?: string | null
}

export interface DishDraftDetail {
  id: number
  name?: string | null
  coverUrl?: string | null
  prepTime?: number | null
  cookTime?: number | null
  difficulty?: number | null
  note?: string | null
  tagIds: number[]
  cuisineIds: number[]
  ingredients: { ingredientId: number; ingredientName?: string | null; amount?: string | null; unitText?: string | null }[]
  steps: { seq?: number; text: string; images?: string | null }[]
}

/** 草稿列表（本人，更新时间倒序）。 */
export function listDrafts(pageNum = 1, pageSize = 10): Promise<Page<DishDraftItem>> {
  return request<Page<DishDraftItem>>({
    url: '/dish/draft/list',
    method: 'GET',
    data: { pageNum, pageSize },
  })
}

/** 存草稿（body 带 id=更新，无 id=新建）→ 草稿 id。 */
export function saveDraft(body: Partial<DishSavePayload> & { id?: number }): Promise<number> {
  return request<number>({ url: '/dish/draft', method: 'POST', data: body })
}

/** 草稿详情（继续编辑回填）。 */
export function draftDetail(id: number): Promise<DishDraftDetail> {
  return request<DishDraftDetail>({ url: `/dish/draft/${id}`, method: 'GET' })
}

/** 删草稿。 */
export function deleteDraft(id: number): Promise<void> {
  return request({ url: `/dish/draft/${id}`, method: 'DELETE' })
}

File: menu-flutter/lib/models/dish.dart
/// 菜品（对应后端 Dish）。字段对齐小程序 dish/List、dish/Detail 的用法。
class Dish {
  final int id;
  final String name;
  final int? cookTime;
  final int? prepTime;
  final int? difficulty;
  final String? note;
  final String? coverUrl;
  /// 来源名（自己创建/下厨房/美食杰/豆果/抖音…，V49）。
  final String? sourceName;
  /// 第三方来源地址（导入时记录，V49）。
  final String? sourceUrl;
  final num? price;
  /// 菜系/分类/标签名（后端 fillRelNames 回填，列表与详情均返回）。
  final List<String> cuisineNames;
  final List<String> categoryNames;
  final List<String> tagNames;
  /// 做过次数（按当前就餐成员统计，search 时回填，缺省 0）。
  final int cookedCount;

  const Dish({
    required this.id,
    required this.name,
    this.cookTime,
    this.prepTime,
    this.difficulty,
    this.note,
    this.coverUrl,
    this.sourceName,
    this.sourceUrl,
    this.price,
    this.cuisineNames = const [],
    this.categoryNames = const [],
    this.tagNames = const [],
    this.cookedCount = 0,
  });

  factory Dish.fromJson(Map<String, dynamic> j) => Dish(
        id: (j['id'] as num).toInt(),
        name: (j['name'] ?? '') as String,
        cookTime: (j['cookTime'] as num?)?.toInt(),
        prepTime: (j['prepTime'] as num?)?.toInt(),
        difficulty: (j['difficulty'] as num?)?.toInt(),
        note: j['note'] as String?,
        coverUrl: j['coverUrl'] as String?,
        sourceName: j['sourceName'] as String?,
        sourceUrl: j['sourceUrl'] as String?,
        price: j['price'] as num?,
        cuisineNames: (j['cuisineNames'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        categoryNames: (j['categoryNames'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        tagNames:
            (j['tagNames'] as List?)?.map((e) => e.toString()).toList() ?? const [],
        cookedCount: (j['cookedCount'] as num?)?.toInt() ?? 0,
      );
}

/// 做法步骤（对应后端 DishStep）。
class DishStep {
  final int? seq;
  final String text;
  final String? images; // 逗号分隔的多图相对路径

  const DishStep({this.seq, required this.text, this.images});

  factory DishStep.fromJson(Map<String, dynamic> j) => DishStep(
        seq: (j['seq'] as num?)?.toInt(),
        text: (j['text'] ?? '') as String,
        images: j['images'] as String?,
      );

  /// 步骤图列表（逗号分隔 → List）。
  List<String> get imageList => images == null || images!.isEmpty
      ? const []
      : images!.split(',').where((s) => s.trim().isNotEmpty).toList();
}

/// 菜品详情聚合（后端 DishDetail record：{dish, steps, cuisineIds, ..., ingredients}）。
class DishDetail {
  final Dish dish;
  final List<DishStep> steps;
  /// 用料明细（食材 id/名/用量克数）。详情页「用料」区展示。
  final List<DishIngredient> ingredients;

  const DishDetail({required this.dish, required this.steps, this.ingredients = const []});

  factory DishDetail.fromJson(Map<String, dynamic> j) => DishDetail(
        dish: Dish.fromJson(j['dish'] as Map<String, dynamic>),
        steps: ((j['steps'] ?? const []) as List)
            .map((e) => DishStep.fromJson(e as Map<String, dynamic>))
            .toList(),
        ingredients: ((j['ingredients'] ?? const []) as List)
            .map((e) => DishIngredient.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// 菜品用料项（对应后端 DishIngredient：食材 id/名/用量克数）。
class DishIngredient {
  final int ingredientId;
  final String? ingredientName;
  /// 用量数量（对应单位个数，如 2 表示 2 个/2 把）。
  final double? amount;
  /// 自然单位名（后端按 unitId 回填；含「适量/少许/一小把」量词单位，§16.3）。
  final String? unitName;
  final double grams;
  /// 库存档位 ENOUGH/LOW/NONE（家里：充足/不足/用完；后端批量回填）。
  final String? stockLevel;

  const DishIngredient({
    required this.ingredientId,
    this.ingredientName,
    this.amount,
    this.unitName,
    this.grams = 0,
    this.stockLevel,
  });

  factory DishIngredient.fromJson(Map<String, dynamic> j) => DishIngredient(
        ingredientId: (j['ingredientId'] as num?)?.toInt() ?? 0,
        ingredientName: j['ingredientName'] as String?,
        amount: (j['amount'] as num?)?.toDouble(),
        unitName: j['unitName'] as String?,
        grams: (j['grams'] as num?)?.toDouble() ?? 0,
        stockLevel: j['stockLevel'] as String?,
      );

  String get displayName => ingredientName ?? '#$ingredientId';

  /// 用量文案：自然单位优先（「2 个」「适量」）；无单位信息回落克数。
  String get amountText {
    final un = unitName;
    if (un != null && un.isNotEmpty) {
      final a = amount;
      if (a != null && a > 0) {
        // 整数不带小数
        final n = a == a.roundToDouble() ? '${a.toInt()}' : a.toStringAsFixed(1);
        return '$n $un';
      }
      return un; // 适量 / 少许（amount 为空）
    }
    final g = grams;
    if (g == g.roundToDouble()) return '${g.toInt()} g';
    return '${g.toStringAsFixed(1)} g';
  }
}
